### PR TITLE
Node annotations

### PR DESCRIPTION
**Pull request by hackaugusto**

### CODE REVIEW

File: raiden/transfer/node.py [hunk 1]
```diff
@@ -732,156 +747,191 @@ def handle_update_transport_authdata(
...
 
 
-def handle_state_change(chain_state: ChainState, state_change: StateChange) -> TransitionResult:
+def handle_state_change(
+        chain_state: ChainState,
+        state_change: StateChange,
+) -> TransitionResult[ChainState]:
+
     if type(state_change) == Block:
+        assert isinstance(state_change, Block), MYPY_ANNOTATION
```
Comment (LefterisJP) [inline raiden/transfer/node.py]: Hmm but why is the assert needed here and in all the files below? Shouldn't the `if type(state_change) == X:` be enough?

File: raiden/transfer/state.py [hunk 2]
```diff
@@ -528,7 +528,7 @@ def from_dict(cls, data: Dict[str, Any]) -> 'TokenNetworkState':
             serialization.identity,
             data['partneraddresses_to_channelidentifiers'],
         )
-        restored.partneraddresses_to_channelidentifiers = defaultdict(
+        restored.partneraddresses_to_channelidentifiers = defaultdict(  # type: ignore
```
Comment (LefterisJP) [inline raiden/transfer/state.py]: What was this complaining about?

File: raiden/transfer/architecture.py [hunk 3]
```diff
@@ -282,7 +284,10 @@ def __ne__(self, other):
         return not self.__eq__(other)
 
 
-class TransitionResult:
+ST = TypeVar('ST', bound=State)
+
+
+class TransitionResult(Generic[ST]):  # pylint: disable=unsubscriptable-object
```
Comment (LefterisJP) [inline raiden/transfer/architecture.py]: Nice use of generic. Note that at the next release of pylint we can remove this `disable` pragma. This is a bug on the pylint side: https://github.com/PyCQA/pylint/issues/2416

File: raiden/transfer/node.py [hunk 4]
```diff
@@ -732,156 +747,191 @@ def handle_update_transport_authdata(
...
             state_change,
         )
     elif type(state_change) == ReceiveLockExpired:
+        assert isinstance(state_change, ReceiveLockExpired), MYPY_ANNOTATION
         iteration = handle_receive_lock_expired(
             chain_state,
             state_change,
         )
 
+    assert chain_state is not None, 'chain_state must be set'
```
Comment (LefterisJP) [inline raiden/transfer/node.py]: @hackaugusto There is a case where the `chain_state` will be `None` here and that is at the very beginning when we process the `ActionInitChain` state change if there is no snapshot.

**Comment by LefterisJP** [comment]
@hackaugusto Have a look at this when you can. Build is now green.

File: raiden/transfer/node.py [hunk 5]
```diff
@@ -765,6 +765,7 @@ def handle_state_change(
         )
     elif type(state_change) == ActionInitChain:
         assert isinstance(state_change, ActionInitChain), MYPY_ANNOTATION
+        at_init = True
```
Comment (rakanalh) [inline raiden/transfer/node.py]: So `chain_state` the paramter at this point is `None`. Wouldn't it be cleaner to set `chain_state = iteration.new_state` after the `handle_chain_init`? we won't need the `at_init` variable in this case

Comment (LefterisJP) [reply raiden/transfer/node.py]: hm indeed. I think you are right. I will amend that commit.

File: raiden/transfer/node.py (hunk 1 above)
Comment (hackaugusto) [reply raiden/transfer/node.py]: It should, but it isn't. mypy does not detect it.

File: raiden/transfer/state.py (hunk 2 above)
Comment (hackaugusto) [reply raiden/transfer/state.py]: type mismatch. defaultdict is not a generic.

File: raiden/transfer/node.py (hunk 1 above)
Comment (LefterisJP) [reply raiden/transfer/node.py]: That's funny. I guess shortcomings of mypy they will hopefully eventually fix. Or ... could it be a subtle different of `type()` vs `isinstance()`? I am just theorizing here. Did not test this.

**Comment by LefterisJP** [comment]
@rakanalh @LefterisJP yeah yeah sorry guys, forgot to approve since I was the last commiter in the branch. Apologies.

File: raiden/transfer/node.py (hunk 1 above)
Comment (LefterisJP) [reply raiden/transfer/node.py]: Hmm actually my theory seems to be correct. Let me open a PR for this.

This only throws warning for the `type()` check:

```python
from typing import Union


class Boo(object):
    a = 1


class Foo(object):
    b = 2


def fun(f: Union[Foo, Boo]):
    if type(f) == Foo:
        print(f.b)

    if isinstance(f, Boo):
        print(f.a)
```

**Pull request merged**